### PR TITLE
misc: add gRPC definition generator

### PR DESCRIPTION
**Pull request by isidentical**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30
* #29
* #28
* #27
* #25
* #23
* #22
* __->__ #21
* #20
* #19

### CODE REVIEW

File: tools/regen_grpc.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class FixGRPCImports(Rule):
+    """Change all unqualified imports to qualified imports."""
+
+    def match(self, node: ast.AST) -> actions.Replace:
+        # import *_pb2
+        assert isinstance(node, ast.Import)
+        assert len(node.names) == 1
+        assert node.names[0].name.endswith("_pb2")
```
Comment (mederka) [inline tools/regen_grpc.py:19]: Should we have a message for when these assertions fail?

File: tools/regen_grpc.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        # from <actual_pkg> import *_pb2
+        parent_dir = self.context.file.resolve().relative_to(PROJECT_ROOT).parent
+        qualified_name = ".".join(parent_dir.parts)
+        return actions.Replace(
+            node,
+            ast.ImportFrom(module=qualified_name, names=node.names, level=0),
+        )
```
Comment (mederka) [inline tools/regen_grpc.py:27]: Neat! I think different unit tests for this would be great.

**Review by mederka — COMMENTED** [review]
This is an interesting workaround, @isidentical. Can we add some tests to make sure it always works as designed?

**Comment by isidentical** [comment]
> This is an interesting workaround, @isidentical. Can we add some tests to make sure it always works as designed?

Thanks for the review @mederka. I think that'd be great (I'll amend some at the #30, which actually touches to this a bit). Just to emphasize, this part will not run with the isolate server, it is merely a tool for us to generate gRPC definitions so I guess it is also something we can improve upon when we start changing the gRPC definitions 😄 But thanks, will try to look into that.

**Pull request merged**